RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc

```diff
@@ -358,38 +545,256 @@ void TrackExtenderWithMTDT<TrackCollection>::produce(edm::Event& ev, const edm::
+  fillValueMap(ev, tracksH, assocOrigTrkRaw, assocOrigTrkName);
 }
 
 namespace {
   bool cmp_for_detset(const unsigned one, const unsigned two) { return one < two; };
 
+  bool getTrajectoryStateClosestToBeamLine(const Trajectory& traj,
+                                           const reco::BeamSpot& bs,
+                                           const Propagator* thePropagator,
+                                           TrajectoryStateClosestToBeamLine& tscbl) {
+    // get the state closest to the beamline
+    TrajectoryStateOnSurface stateForProjectionToBeamLineOnSurface =
+        traj.closestMeasurement(GlobalPoint(bs.x0(), bs.y0(), bs.z0())).updatedState();
+
+    if (!stateForProjectionToBeamLineOnSurface.isValid()) {
+      edm::LogError("CannotPropagateToBeamLine") << "the state on the closest measurement isnot valid. skipping track.";
+      return false;
+    }
+
+    const FreeTrajectoryState& stateForProjectionToBeamLine = *stateForProjectionToBeamLineOnSurface.freeState();
+
+    TSCBLBuilderWithPropagator tscblBuilder(*thePropagator);
+    tscbl = tscblBuilder(stateForProjectionToBeamLine, bs);
+
+    return tscbl.isValid();
+  }
+
+  bool trackPathLength(const Trajectory& traj,
+                       const reco::BeamSpot& bs,
+                       const Propagator* thePropagator,
+                       double& pathlength) {
+    pathlength = 0.;
+
+    TrajectoryStateClosestToBeamLine tscbl;
+    bool tscbl_status = getTrajectoryStateClosestToBeamLine(traj, bs, thePropagator, tscbl);
+
+    if (!tscbl_status)
+      return false;
+
+    bool validpropagation = true;
+    double pathlength1 = 0.;
+    double pathlength2 = 0.;
+
+    //add pathlength layer by layer
+    for (auto it = traj.measurements().begin(); it != traj.measurements().end() - 1; ++it) {
+      const auto& propresult = thePropagator->propagateWithPath(it->updatedState(), (it + 1)->updatedState().surface());
+      double layerpathlength = std::abs(propresult.second);
+      if (layerpathlength == 0.) {
+        validpropagation = false;
+      }
+      pathlength1 += layerpathlength;
+    }
+
+    //add distance from bs to first measurement
+    if (traj.direction() == alongMomentum) {
+      const auto& propresult2 =
+          thePropagator->propagateWithPath(tscbl.trackStateAtPCA(), traj.firstMeasurement().updatedState().surface());
+      pathlength2 = propresult2.second;
+      if (pathlength2 == 0.) {
+        validpropagation = false;
+      }
+      pathlength = pathlength1 + pathlength2;
+    } else {
+      const auto& propresult2 =
+          thePropagator->propagateWithPath(tscbl.trackStateAtPCA(), traj.lastMeasurement().updatedState().surface());
+      pathlength2 = propresult2.second;
+      if (pathlength2 == 0.) {
+        validpropagation = false;
+      }
+      pathlength = pathlength1 + pathlength2;
+    }
+
+    return validpropagation;
+  }
+
+  struct TrackTofPidInfo {
+    TrackTofPidInfo(double magp2,
```

Comment:
this struct exposes all data members which are supposed to have some relationship.
This is a case for making a class.
Perhaps a more effective way out here is to keep a plain ```struct TrackTofPidInfo ``` without the constructor and replace the constructor with a standalone function that returns a ```const TrackTofPidInfo```.